python-checks/src/main/java/org/sonar/python/checks/SklearnPipelineParameterAreCorrectCheck.java

```diff
@@ -0,0 +1,320 @@
+        }
+      });
+    }
+  }
+
+  private static void createIssue(SubscriptionContext subscriptionContext, String parameter, Set<StringAndTree> stringAndTree) {
+    stringAndTree.stream().filter(stringAndTree1 -> stringAndTree1.string().equals(parameter)).findFirst()
+      .ifPresent(location -> subscriptionContext.addIssue(location.tree, MESSAGE));
+  }
+
+  private static boolean isNotAValidParameter(String parameter, List<FunctionSymbol.Parameter> possibleParameters) {
+    return possibleParameters.stream().noneMatch(symbol -> Objects.equals(symbol.name(), parameter));
+  }
+
+  private static Optional<FunctionSymbol> getInitFunctionSymbol(ClassSymbol classSymbol) {
+    return classSymbol.declaredMembers().stream().filter(
+      memberSymbol -> "__init__".equals(memberSymbol.name())).findFirst().map(FunctionSymbol.class::cast);
+  }
+
+  private static Stream<Expression> getExpressionsFromArgument(@Nullable RegularArgument argument) {
+    return Optional.ofNullable(argument)
+      .map(RegularArgument::expression)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(ListLiteralImpl.class))
+      .map(ListLiteralImpl::elements)
+      .map(ExpressionList::expressions)
+      .stream()
+      .flatMap(Collection::stream);
+  }
+
+  private record ExpressionAndPrefix(List<Expression> tuple, String prefix) {
+  }
+
+  private static Map<String, ClassSymbol> parsePipeline(CallExpression callExpression) {
+    var stepsArgument = TreeUtils.nthArgumentOrKeyword(0, "steps", callExpression.arguments());
+    var out = new HashMap<String, ClassSymbol>();
+
+    var a = getExpressionsFromArgument(stepsArgument)
+      .map(
+        TreeUtils.toInstanceOfMapper(TupleImpl.class))
+      .filter(Objects::nonNull)
+      .map(TupleImpl::elements)
+      .filter(SklearnPipelineParameterAreCorrectCheck::isTwoElementTuple)
+      // If we find a pipeline inside the pipeline, we need to parse it recursively
+      .map(SklearnPipelineParameterAreCorrectCheck::createEmptyExpressionAndPrefix)
+      .flatMap(expandRecursivePipelines());
+    a.forEach(
+      expressionAndPrefix -> getResult(expressionAndPrefix.tuple()).ifPresent(result1 -> out.put(expressionAndPrefix.prefix() + result1.stepName(), result1.classifierName())));
+    return out;
+  }
+
+  private static boolean isTwoElementTuple(List<Expression> elements) {
+    return elements.size() == 2;
+  }
+
+  private static ExpressionAndPrefix createEmptyExpressionAndPrefix(List<Expression> tuple) {
+    return new ExpressionAndPrefix(tuple, "");
+  }
+
+  private static Function<ExpressionAndPrefix, Stream<ExpressionAndPrefix>> expandRecursivePipelines() {
+    return expressionAndPrefix -> {
+      var tuple = expressionAndPrefix.tuple();
+      var step = tuple.get(0);
+      var classifier = tuple.get(1);
+
+      if (!step.is(Tree.Kind.STRING_LITERAL)) {
+        return Stream.of(expressionAndPrefix);
+      }
+
+      if (classifier.is(Tree.Kind.NAME)) {
```

Comment:
I guess we could merge these ifs together. As the code is quite complex, having less branching to go through make it easier to reason about.